### PR TITLE
[BugFix] Make Stage Device Initialization Respect Visible GPUs

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
## Purpose
See details in https://github.com/vllm-project/vllm-omni/issues/2024; this handles the mapping, cleans up the code, and also fixes the log to show the correct devices.

- Also fixes diffusion stage polluting the visible device vars, as this will now cause sp tests to fail.

## Test Plan
- Added a test ensuring that the out of bounds case is caught

## Test Result
Logs show the correct physical devices now when we need to map:
```bash
 CUDA_VISIBLE_DEVICES=2,3,4,6  VLLM_LOGGING_LEVEL=debug vllm serve   black-forest-labs/FLUX.2-klein-4B --omni  --cache-backend 
cache_dit --cfg-parallel-size 2 --ulysses-degree 2
```

```
(APIServer pid=544872) INFO 03-19 22:18:23 [stage_init_utils.py:222] [stage_init] Stage-0 set runtime devices: 2,3,4,6
```

We raise quickly if we don't have enough GPUs and CUDA_VISIBLE_DEVICES is directly set,
```bash
CUDA_VISIBLE_DEVICES=2,3 VLLM_LOGGING_LEVEL=debug vllm serve   black-forest-labs/FLUX.2-klein-4B --omni  --cache-backend cache_dit --cfg-parallel-size 2 --ulysses-degree 2
...
(APIServer pid=543361) RuntimeError: Orchestrator initialization failed: Stage 0 requires 4 devices, but only 2 devices are visible
```

And the test passes.

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
@linyueqian @tzhouam PTAL

File: vllm_omni/entrypoints/stage_utils.py [hunk 1]
```diff
@@ -6,115 +6,103 @@
...
+    if num_visible < num_logical:
+        raise ValueError(f"Stage {stage_id} requires {num_logical} devices, but only {num_visible} devices are visible")
+
+    # Ensure that the logical IDs are actually in range to avoid index errors;
+    # If the check above passes and those below fail, the logical devices are wrong,
+    # i.e., not actually 0, 1, ..., n
+    if not all(device.isdigit() for device in device_list):
+        raise ValueError("Logical devices must be non-negative integers")
+
+    logical_ids = [int(device) for device in device_list]
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/stage_utils.py:100]: Nit: `min(logical_ids) < 0` is dead code — `isdigit()` on L95 already rejects anything with a leading `-`.

File: vllm_omni/entrypoints/stage_utils.py [hunk 2]
```diff
@@ -6,115 +6,103 @@
...
+            raise ValueError("Device IDs must be non-negative integers!")
+        return [str(devices)]
+    # Devices will usually be ints, but not always
+    # so we don't explicitly validate that here.
+    return [t.strip() for t in devices.split(",") if t.strip() != ""]
+
+
+def _map_device_list(stage_id: int, device_list: list[str], visible_device_list: list[str]):
+    """Maps logical to physical devices if we have enough visible devices available.
+
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/stage_utils.py:82]: Missing return type annotation.
```suggestion
def _map_device_list(stage_id: int, device_list: list[str], visible_device_list: list[str]) -> list[str]:
```

File: vllm_omni/entrypoints/stage_utils.py (hunk 1 above)
Comment (alex-jw-brooks) [reply vllm_omni/entrypoints/stage_utils.py:100]: removed 🙂

**Comment by yinpeiqi** [comment]
LGTM

**Comment by hsliuustc0106** [comment]
## 🔍 Code Review

Excellent refactoring! The code is much cleaner and the device mapping bug fix is critical. A few minor suggestions:

### 🔧 Minor Issues

#### 1. Docstring Bug 🔴
The `_map_device_list` docstring has a duplicate function signature inside it:
```python
"""Maps logical to physical devices...
def _map_device_list(stage_id: int, device_list: list[str], visible_device_list: list[str]) -> list[str]:
    Args:
```
This looks like an editing artifact - should be removed.

#### 2. Missing Type Hints 🟡
```python
def _map_device_list(stage_id: int, device_list: list[str], 
                     visible_device_list: list[str]):  # <- missing return type
```
Should be:
```python
def _map_device_list(...) -> list[str]:
```

#### 3. Test Coverage 🟡
Current test is great for the "not enough devices" case. Consider adding:
- Invalid logical IDs (negative numbers, non-digits)
- Single device case (both int and str)
- CPU device case (devices=None or "cpu")
- Edge case where `max(logical_ids) >= num_visible`

### 💡 Suggestions

The `device.isdigit()` check is good for logical devices (which should always be 0, 1, 2, ...). Consider adding a comment to clarify this assumption:
```python
# Logical devices are always 0-indexed integers
# (unlike physical devices which may have arbitrary IDs)
if not all(device.isdigit() for device in device_list):
    raise ValueError("Logical devices must be non-negative integers")
```

### ✅ Strengths

- 🎯 Fixes critical device mapping bug
- 🧹 Clean separation of concerns with helper functions
- 🔥 Better error handling (fail fast instead of silent warnings)
- 📊 Improved logging with physical device IDs
- ✅ Good test for the "not enough devices" case

**Overall**: Solid bugfix with great code cleanup! Just the minor docstring issue and a few type hints to address. After fixing the docstring, this is ready to merge. 👍

**Review by linyueqian — APPROVED** [review]
LGTM

**Comment by linyueqian** [comment]
fix ci please

**Comment by alex-jw-brooks** [comment]
Thanks @linyueqian @lishunyang12. The CI was failing because the diffusion stage (which runs in the same process) was polluting CUDA_VISIBLE_DEVICES. I went ahead and fixed it as part of this PR as well since it's related
